### PR TITLE
Unified Cocaine ACL Tooling

### PR DESCRIPTION
**Pull request by karitra**
## Unified Cocaine ACL Tooling (UNICAT) plugin

Service for unified ACL records altering.

Preliminary implementation, ready to review.

**TODO:**
- [x] storage backend
- [x] unicorn backend
- [x] unicorn value version support!
- [x] complete grant/revoke pair for storage, unicorn
- [ ] ensure correct Unicorn prefix access semantics
- [x] ~~find better place for common stuff: `metainfo_t`, `indentity_t`, `flags_t`~~ (separate PR)
- [x] ~~tooling integration~~ (separate PR in tools)
- [ ] ensure correct cid/uid access restriction for grant/revoke (unitests?)
- [x] ~~additional methods: create/view/reset~~ (separate PR) 
- [x] ~~unit tests~~ (separate PR)

### CODE REVIEW

**Comment by karitra** [comment]
Yet lacks functional tests and marginal cases testing, but @antmat & @3Hren PTAL.

**Comment by karitra** [comment]
Yet lacks functional tests and marginal cases testing, but @antmat & @3Hren PTAL.

**UPD2**: following issue hopefully solved in 68f6e52.
**UPD1**: oops, testing results for unicorn, please standby:

```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::t
hread_resource_error> >'
  what():  boost thread: trying joining itself: Resource deadlock avoided
```

**Comment by karitra** [comment]
@antmat, @3Hren, pls, don't beat me up for such a huge pile of garbage, but it seems the only way to animate all those async stuff, hopefully it will be party refactored out later, so PTAL.

File: unicat/include/cocaine/auth/metainfo.hpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        std::map<auth::cid_t, auth::flags_t>,
+        std::map<auth::uid_t, auth::flags_t>
+    >;
+
+    static
+    result_type
+    convert(const dynamic_t& from) {
+        auto& tuple = from.as_array();
+        if (tuple.size() != 2) {
+            throw std::bad_cast();
```
Comment (3Hren) [inline unicat/include/cocaine/auth/metainfo.hpp]: It's really misleading exception type. Consider `std::invalid_argument` instead.

File: unicat/include/cocaine/detail/forwards.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#pragma once
```
Comment (3Hren) [inline unicat/include/cocaine/detail/forwards.hpp]: Avoid `detail` section, especially in plugins, move in `src` instead.

File: unicat/src/backend/backend.hpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+namespace cocaine { namespace unicat {
+
+using version_t = unicorn::version_t;
+
+namespace async {
+
+// Used in unicorn backend reincarnation for scope management.
+struct scope_ctl_t {
+    auto attach_scope(api::unicorn_scope_ptr s) -> void {
+        scope = s;
```
Comment (3Hren) [inline unicat/src/backend/backend.hpp]: What to do with previous scope?

File: unicat/src/backend/unicorn.cpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    }
+}
+
+unicorn_backend_t::unicorn_backend_t(const options_t& options) :
+    backend_t(options),
+    access(options.access_handlers_cache->make_handler<api::authorization::unicorn_t>(options.name)),
+    backend(api::unicorn(options.ctx_ref, options.name))
+{
+    COCAINE_LOG_DEBUG(this->logger(), "unicat::unicorn backend started '{}'", this->get_options().name);
+    dbg("unicorn backend is alive");
```
Comment (3Hren) [inline unicat/src/backend/unicorn.cpp]: Do we need these logs?

File: unicat/src/backend/unicorn.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+unicorn_backend_t::~unicorn_backend_t()
+{
+    COCAINE_LOG_DEBUG(this->logger(), "unicat::unicorn backend detached '{}'", this->get_options().name);
+    dbg("unicorn backend is dead");
```
Comment (3Hren) [inline unicat/src/backend/unicorn.cpp]: And this.

File: unicat/src/completion.hpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            // TODO: make errors with <code, string> pairs
+            std::vector<std::string> errors;
+
+            for(auto& comp : state) {
+                // TODO: better error case separation: exception or error_code
+                //       by authorization::verify handler
+                if (comp.has_error()) {
+                    if (comp.has_error_code()) {
+                        errors.push_back(comp.make_access_error());
+                        continue;
```
Comment (3Hren) [inline unicat/src/completion.hpp]: else?

File: unicat/src/metainfo.cpp [hunk 7]
```diff
@@ -0,0 +1,150 @@
+#include <boost/range/algorithm/transform.hpp>
+
+#include "cocaine/auth/metainfo.hpp"
+#include "cocaine/idl/unicat.hpp"
+
+#if 0
```
Comment (3Hren) [inline unicat/src/metainfo.cpp]: Drop this doubtful stuff.

File: unicat/src/module.cpp [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#include "cocaine/service/unicat.hpp"
+
+#include <cocaine/repository.hpp>
+#include <cocaine/repository/service.hpp>
+
+extern "C" {
+    auto validation() -> cocaine::api::preconditions_t {
+        return cocaine::api::preconditions_t{COCAINE_MAKE_VERSION(0, 12, 0)};
```
Comment (3Hren) [inline unicat/src/module.cpp]: 12.14

File: unicat/src/unicat.cpp [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+    {
+        std::string scheme;
+        std::string entity;
+        boost::optional<std::string> service;
+
+        entities_by_scheme_type separated;
+
+        for(const auto& el : entities) {
+            std::tie(scheme, service, entity) = el;
+            separated[cu::scheme_from_string(scheme)]
```
Comment (3Hren) [inline unicat/src/unicat.cpp]: Simplify.

**Comment by 3Hren** [comment]
At last:

```
Undefined symbols for architecture x86_64:
  "_msgpack_unpack", referenced from:
      void cocaine::api::storage_t::get<cocaine::auth::metainfo_t>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<void (std::__1::future<cocaine::auth::metainfo_t>)>)::'lambda'(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >)::operator()(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >) const in storage.cpp.o
  "_msgpack_zone_destroy", referenced from:
      void cocaine::api::storage_t::get<cocaine::auth::metainfo_t>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<void (std::__1::future<cocaine::auth::metainfo_t>)>)::'lambda'(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >)::operator()(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >) const in storage.cpp.o
  "_msgpack_zone_init", referenced from:
      void cocaine::api::storage_t::get<cocaine::auth::metainfo_t>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<void (std::__1::future<cocaine::auth::metainfo_t>)>)::'lambda'(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >)::operator()(std::__1::future<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >) const in storage.cpp.o
```

**Review by 3Hren — CHANGES_REQUESTED** [review]

**Comment by karitra** [comment]
@3Hren PTAL.

File: unicat/src/unicat.cpp [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+                errors.push_back(std::current_exception());
+            }
+        }
+        return errors;
+    }
+} // detail
+
+namespace dbg {
+    auto
+    log_requested_entities(std::shared_ptr<logging::logger_t> log, const std::vector<cu::entity_type>& entities)
```
Comment (3Hren) [inline unicat/src/unicat.cpp:115]: Seems like dead code.

Comment (karitra) [reply unicat/src/unicat.cpp:115]: Yep, but it'll come handy in case of debug, should it be removed?

**Comment by karitra** [comment]
@3Hren, PTAL.

**Review by 3Hren — APPROVED** [review]

**Comment by karitra** [comment]
Need bit more testing, probably unit-tests for metainfo alter cases + test for unicorn path, more intensive load test, exceptional cases test, but ready to be merged in order to be able to move further.

**Pull request merged**